Added tooltip in aes using tippy.js

suhaani-agarwal opened the pull request:
Resolves issue #179 

I have added the tooltip parameter for aes and this is the initial PR for the same. 
This is the video demonstration of the output that I got on rendering a viz using this tooltip implementation:
https://youtu.be/gAa8a4jaKMk

- Right now the index.html contains definite links but I plan to make them relative like the others inside vendor/ 

- If you approve this, I will write a test for this tooltip implementation.

tdhock commented:
this looks good, thanks for the contribution!
yes please fix existing tests and add new tests. 
Do you know how it work? How can it be tested?
The testing for the previous tooltip implementation was easy (just look for <title> element in HTML).

tdhock commented on R/geom-text.r:
```diff
@@ -126,7 +126,8 @@ geom_text <- function(mapping = NULL, data = NULL,
                       show.legend = NA,
-                      inherit.aes = TRUE)
+                      inherit.aes = TRUE,
+                      tooltip = NULL)
```
I don't believe this should be necessary. Can you please undo?

suhaani-agarwal replied:
> I don't believe this should be necessary. Can you please undo?

If I am removing `tooltip = NULL` , its not recognising tooltip as a parameter 
It shows 
`Error in geom_text(aes(life.expectancy, fertility.rate, label = country,  : 
  object 'tooltip' not found
Calls: geom_text -> layer
Execution halted`

suhaani-agarwal commented:
> this looks good, thanks for the contribution! yes please fix existing tests and add new tests. Do you know how it work? How can it be tested? The testing for the previous tooltip implementation was easy (just look for <title> element in HTML).

Hi sorry for the late reply, I was travelling. Yes I will write the tests according to the existing tests....

tdhock commented on inst/htmljs/animint.js:
```diff
@@ -1061,6 +1061,7 @@ var animint = function (to_select, json_file) {
     var panel_g_element = layer_g_element.select("g.PANEL" + PANEL);
     var elements = panel_g_element.selectAll(".geom");
 
+    
```
please undo addition of empty lines

tdhock commented on inst/htmljs/animint.js:
```diff
@@ -1609,29 +1610,50 @@ var animint = function (to_select, json_file) {
+          return d.tooltip;
+        };
+      } else if (has_clickSelects) {
+        // Using the default tooltip text for clickSelects
+        tooltipText = function (d) {
           var v_name = g_info.aes.clickSelects;
           return v_name + " " + d.clickSelects;
-	};
-      }else{ //clickSelects_variable
-	text_fun = function(d){
-	  return d["clickSelects.variable"] + " " + d["clickSelects.value"];
-	};
-      }
-      // if elements have an existing title, remove it.
+        };
+      } else { // clickSelects_variable
+        // Using the default tooltip text for clickSelects.variable
+        tooltipText = function (d) {
+          return d["clickSelects.variable"] + " " + d["clickSelects.value"];
+        };
+      }
```
here it looks like you just changed text_fun to tooltipText, but there are lots of other lines changed too.
please avoiding changing lines that are not necessary.
can you please undo all of these changes, and instead use text_fun as in the original code?

tdhock replied:
click Files Changed tab in github to see what changes will appear for me to review

suhaani-agarwal commented:
> that's a good question, I'm not sure how to simulate hovering over a component. it looks like these people did something similar, [rstudio/chromote#92](https://github.com/rstudio/chromote/issues/92) can you please try sometihng like that?

Thankyou for this, it helped a lot. I have written two tests for this tooltip implementation.`test-renderer-tippytooltip`  and  ` test-renderer-tippy-tooltip-interactivity.r` . These are the versions of the old tests : `test-renderer1-tooltip` and `test-renderer3-tooltip-interactivity` made for the new implementation of tooltip. 
If these changes are okay, shall I rename the new test files as the old ones?
Also, should I make a demonstration video for these tests?

tdhock commented on inst/htmljs/animint.js:
```diff
@@ -1626,11 +1627,22 @@ var animint = function (to_select, json_file) {
       }
-      // if elements have an existing title, remove it.
+      // Setting the data-tippy-content attribute on each element
+      elements.each(function (d) {
```
please try elements.attr(text_fun) instead

(notice that .each is not used anywhere else in code)

tdhock commented:
also please modify the tests in test-renderer1-tooltip.R

tdhock commented on tests/testthat/test-renderer-tippy-tooltip-interactivity.r:
```diff
@@ -0,0 +1,53 @@
+  rect_center_y <- rect_y + rect_height / 2
+  
+  # Simulate hover on rectangle
+  remDr$Input$dispatchMouseEvent(type = "mouseMoved", x = rect_center_x, y = rect_center_y)
```
if this does not work for some reason, we can change to testing the data-tippy-content attribute

suhaani-agarwal replied:
Hi, I run the test at my end and it was working fine and the tests were passing. 
It was simulating hovering over the rectangle component and checking that aria-expanded became = true after the mouse hover correctly.

suhaani-agarwal commented:
- I have modified test-renderer1-tooltip.R and test-renderer3-tooltip-interactivity.R according to new tooltip implementation. 
-  I have also modified the tooltip tests in two other files :
test-renderer1-variable-value.R and test-renderer3-prostateLasso.R
- I have run these tests and they are passing successfully

tdhock commented:
tests are not passing on CI, please fix

tdhock commented:
I don't know how to fix, sorry.
but we definitely would need to fix this before merging.
maybe file an issue with tippy or chromote to ask for help?

suhaani-agarwal commented:
> I don't know how to fix, sorry.
> 
> but we definitely would need to fix this before merging.
> 
> maybe file an issue with tippy or chromote to ask for help?

Okay, I will work towards fixing it!

tdhock commented on tests/testthat/test-renderer3-tooltip-interactivity.R:
```diff
@@ -4,56 +4,47 @@ data("CO2")
+  # Initial state - aria-expanded should be "false"
+  initial_circle <- getNodeSet(info$html, '//circle[@data-tippy-content="Mn3"]')[[1]]
+  initial_rect <- getNodeSet(info$html, '//rect[@data-tippy-content="Test Rectangle"]')[[1]]
+ 
```
please undo addition of empty lines inside test_that blocks
